perf/memtx.cc

```diff
@@ -0,0 +1,199 @@
+					    &field_def_default, 0);
+		if (space_def == nullptr)
+			panic("failed to create new space definition");
+		RLIST_HEAD(key_list);
+		struct index_opts idx_opts;
+		::index_opts_create(&idx_opts);
+		std::uint32_t fields[] = { 0 };
+		std::uint32_t types[] = { FIELD_TYPE_UNSIGNED };
+		key_def = ::box_key_def_new(fields, types, 1);
+		if (key_def == nullptr)
+			panic("failed to create new key definition");
+		tree_idx_def = ::index_def_new(sid, tree_iid, "pk",
+					       std::strlen("pk"), TREE,
+					       &idx_opts, key_def, nullptr);
+		if (tree_idx_def == nullptr)
+			panic("failed to create new index definition");
+		::index_def_list_add(&key_list, tree_idx_def);
+		space = ::memtx_space_new(memtx, space_def, &key_list);
+		if (space == nullptr)
+			panic("failed to create new space");
+		::space_cache_replace(nullptr, space);
+		if (::space_add_primary_key(space) != 0)
+			panic("failed to add primary key");
+
+		for (auto &k: key_set)
+			::box_insert(sid, k.first, k.second, nullptr);
+	}
+
+	void
+	TearDown(const benchmark::State &) final
+	{
+		::space_def_delete(space_def);
+		::key_def_delete(key_def);
+
+		::session_free();
+		::user_cache_free();
+		::space_cache_destroy();
+		::tuple_free();
+
+		memtx->base.vtab->shutdown(&memtx->base);
+
+		::fiber_free();
+		::memory_free();
+	}
+
+private:
+	void
+	generate_key_set()
+	{
+		for (std::size_t i = 0; i < key_set.size(); ++i) {
+			key_set[i].second =
+				::mp_encode_array(key_set[i].first, 1);
+			key_set[i].second =
+				::mp_encode_uint(key_set[i].second, i);
+		}
+	}
+
+	static constexpr unsigned int rand_seed = 0;
+
+	static constexpr std::uint32_t memtx_objsize_min = 16;
+	static constexpr unsigned int memtx_granularity = 8;
+	static constexpr float memtx_alloc_factor = 1.1;
+	struct memtx_engine *memtx{};
+
+	struct space_def *space_def{};
+	struct key_def *key_def{};
+	struct index_def *tree_idx_def{};
+	struct space *space{};
+
+	static constexpr std::size_t key_size_max = 5 + 9;
+	static constexpr std::size_t key_set_size = 1 << 10;
+	std::array<std::pair<char[key_size_max], char *>, key_set_size> key_set;
+	static constexpr std::size_t key_subset_size = key_set_size / 2;
+
+protected:
+	void
+	generate_key_subset()
+	{
+		for(auto &k: key_subset)
+			k = key_set[std::rand() % key_set.size()];
+	}
+
+	static constexpr std::uint32_t uid = 0;
+	static constexpr std::uint32_t sid = 0;
+	static constexpr std::uint32_t tree_iid = 0;
+
+	std::array<std::pair<char *, char *>, key_subset_size> key_subset;
+
+	static constexpr std::uint32_t select_lim = 4294967295;
+	static constexpr std::size_t empty_key_size_max = 5;
+	std::pair<char[empty_key_size_max], char *> empty_key;
+};
+
+BENCHMARK_F(MemtxFixture, TreeGetRandomExisting)
+(benchmark::State &state)
+{
+	std::int64_t iterations_count = 0;
+	for (auto _: state) {
+		for (const auto &k: key_subset) {
+			++iterations_count;
+			struct tuple *t;
+			benchmark::DoNotOptimize(::box_index_get(sid, tree_iid,
```

Comment:
I guess we should somehow make sure that index_get is proceeded without errors

Reply:
I guess this can be checked once manually. On the other hand, adding a check probably won't harm. @alyapunov what do you think?

Reply:
That would be good but I think not necessary. I bet in case of errors or nops we will explicitly notice it by bench results.